packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/cluster_client.ts

```diff
@@ -78,28 +78,43 @@ export class ClusterClient implements ICustomClusterClient {
       kibanaVersion,
     });
     this.rootScopedClient = configureClient(config, {
+      scoped: true,
       logger,
       type,
       getExecutionContext,
-      scoped: true,
       agentFactoryProvider,
       kibanaVersion,
     });
   }
 
   asScoped(request: ScopeableRequest) {
-    const scopedHeaders = this.getScopedHeaders(request);
+    const createScopedClient = () => {
+      const scopedHeaders = this.getScopedHeaders(request);
 
-    const transportClass = createTransport({
-      getExecutionContext: this.getExecutionContext,
-      getUnauthorizedErrorHandler: this.createInternalErrorHandlerAccessor(request),
-    });
+      const transportClass = createTransport({
+        getExecutionContext: this.getExecutionContext,
+        getUnauthorizedErrorHandler: this.createInternalErrorHandlerAccessor(request),
+      });
+
+      return this.rootScopedClient.child({
+        headers: scopedHeaders,
+        Transport: transportClass,
+      });
+    };
+
+    const createSecondaryScopedClient = () => {
+      const secondaryAuthHeaders = this.getSecondaryAuthHeaders(request);
+
+      return this.asInternalUser.child({
+        headers: secondaryAuthHeaders,
+      });
+    };
 
-    const scopedClient = this.rootScopedClient.child({
-      headers: scopedHeaders,
-      Transport: transportClass,
+    return new ScopedClusterClient({
+      asInternalUser: this.asInternalUser,
+      asCurrentUserClientFactory: createScopedClient,
+      asSecondaryAuthClientFactory: createSecondaryScopedClient,
```

Comment:
Creating child clients is somewhat costly. Did some quick refactoring to have them lazily instantiated instead of eagerly. (it's especially important now that we have 2 different instantiations that can happen with `asCurrentUser` and `asSecondaryAuth`)